README.md
# tsmaz

A port of the [smaz](https://github.com/antirez/smaz) small string compression library.

From the original library:

> Smaz is a simple compression library suitable for compressing very short
> strings. General purpose compression libraries will build the state needed
> for compressing data dynamically, in order to be able to compress every kind
> of data. This is a very good idea, but not for a specific problem: compressing
> small strings will not work.
>
> Smaz instead is not good for compressing general purpose data, but can compress
> text by 40-50% in the average case (works better with English), and is able to
> perform a bit of compression for HTML and urls as well. The important point is
> that Smaz is able to compress even strings of two or three bytes!
>
> For example the string "the" is compressed into a single byte.
>
> To compare this with other libraries, think that like zlib will usually not be able to compress text shorter than 100 bytes.

## Usage

Install,
```sh
npm install tsmaz
```

then,
```javascript
const { compress, decompress } = require('tsmaz');

const compressed = compress('foobar');
console.log(decompress(compressed));
```

It is also possible to use a custom codebook:
```javascript
const { Smaz } = require('tsmaz');

// NOTE: this array needs to have a maximum length of 254!
const smaz = new Smaz([
  'foo',
  'bar',
  'foobar',
  'er',
  'ab',
  'aa',
];

smaz.decompress(smaz.compress('foo'));
```

## Generating Codebooks

The library exports a `generate` function, which accepts a list of strings which
you would like to compress, then learns a dictionary which works well with the
kind of data given as input.

```javascript
const { generate, Smaz } = require('tsmaz');

// Generate codebook
const codebook = generate([
  'foo',
  'foobar',
  'bar',
  'baz-bar',
]);

// Use custom codebook
const smaz = new Smaz(codebook);

const compressed = smaz.compress('foo-barbaz');
console.log(compressed);
const original = smaz.decompress(compressed);
console.log(original);
```

## Performance

`tsmaz` makes use of a trie data-structure for look-up (whereas the original
implementation used a hashtable). Apart from that the behavior should be the
same, and performance is pretty good (around 32k bytes per millisecond for
compression and 70k bytes per millisecond for decompression).


CHANGELOG.md
# Changelog

### 1.3.0

*2019-09-12*

  - add ability to know size of compressed string without doing the actual compression

### 1.2.2

*2019-06-07*

  - prevent global computation on import and expose both cjs and es6 sources

### 1.2.1

*2019-05-19*

 - Add benchmarks + small optimization [27be341](https://github.com/remusao/tsmaz/commit/27be341fe3a91d36f464d8f91a9e995698a7a293)

### 1.2.0

*2019-05-16*

  - Use optimized bundle in tests + fixes [d253233](https://github.com/remusao/tsmaz/commit/d253233a9db418ee4c88df8aef8dfc95ca73735d)
    * Import optimized cjs bundle + type definitions for tests
    * Change API of factory to not rely on named attributes

### 1.1.0

*2019-05-16*

  - Produce optimized and minified bundles [#1](https://github.com/remusao/tsmaz/pull/1)
    * Add closure compiler plugin to rollup
    * Remove dependency to tslib
    * Produce minified bundles
    * Produce optimized bundles
    * Change default entrypoints in package.json
    * Simplify rollup config
    * Remove un-needed dependencies + update other ones

### 1.0.3

*2019-04-17*

  - Improve codebook generation [26fc0ad](https://github.com/remusao/tsmaz/commit/26fc0addcfebd95ab68afdcb1934e30fb56c847d)
  - Update dev dependencies [e5bdbc9](https://github.com/remusao/tsmaz/commit/e5bdbc9809dadea67c5f430253e54429e22b8974)
  - Harden tsconfig.json [8393224](https://github.com/remusao/tsmaz/commit/8393224f9bf4b69d04ab61f49c5775e691e6f0b4)
  - Restructure code and expose codebook generation from bundle [726e2bb](https://github.com/remusao/tsmaz/commit/726e2bbc1576f2c6122ae7d26f5a9298e513f344)
  - Remove dependency to ts-node [96fc207](https://github.com/remusao/tsmaz/commit/96fc2073dc953453b314442854cab3d531bb246e)
  - Target ES3 instead of ES2018 [bb05d27](https://github.com/remusao/tsmaz/commit/bb05d2774c1adc9d5ea542d13aa84b9297fd58b8)
  - Fix tests warning [799c494](https://github.com/remusao/tsmaz/commit/799c49495d54b9637fd059ae4df0550d24d4d48c)
  - Add more tests [5d2e656](https://github.com/remusao/tsmaz/commit/5d2e656f233a907cb685a1d42e8068219b89491a)

### 1.0.2

*2019-03-18*

  - Add script to generate optimized codebooks

### 1.0.1

*2019-03-17*

  - Initial release


bench.js
const Benchmark = require('benchmark');
const { compress, decompress, factory } = require('.');

(function main() {
  const suite = new Benchmark.Suite;

  suite
    .add('#factory', function() {
      return factory([
        ' ',
        'the',
        'e',
        't',
        'a',
        'of',
        'o',
        'and',
        'i',
        'n',
        's',
        'e ',
        'r',
        ' th',
        ' t',
        'in',
        'he',
        'th',
        'h',
        'he ',
        'to',
        '\r\n',
        'l',
        's ',
        'd',
        ' a',
        'an',
        'er',
        'c',
        ' o',
        'd ',
        'on',
        ' of',
        're',
        'of ',
        't ',
        ', ',
        'is',
        'u',
        'at',
        '   ',
        'n ',
        'or',
        'which',
        'f',
        'm',
        'as',
        'it',
        'that',
        '\n',
        'was',
        'en',
        '  ',
        ' w',
        'es',
        ' an',
        ' i',
        'f ',
        'g',
        'p',
        'nd',
        ' s',
        'nd ',
        'ed ',
        'w',
        'ed',
        'http://',
        'https://',
        'for',
        'te',
        'ing',
        'y ',
        'The',
        ' c',
        'ti',
        'r ',
        'his',
        'st',
        ' in',
        'ar',
        'nt',
        ',',
        ' to',
        'y',
        'ng',
        ' h',
        'with',
        'le',
        'al',
        'to ',
        'b',
        'ou',
        'be',
        'were',
        ' b',
        'se',
        'o ',
        'ent',
        'ha',
        'ng ',
        'their',
        '"',
        'hi',
        'from',
        ' f',
        'in ',
        'de',
        'ion',
        'me',
        'v',
        '.',
        've',
        'all',
        're ',
        'ri',
        'ro',
        'is ',
        'co',
        'f t',
        'are',
        'ea',
        '. ',
        'her',
        ' m',
        'er ',
        ' p',
        'es ',
        'by',
        'they',
        'di',
        'ra',
        'ic',
        'not',
        's, ',
        'd t',
        'at ',
        'ce',
        'la',
        'h ',
        'ne',
        'as ',
        'tio',
        'on ',
        'n t',
        'io',
        'we',
        ' a ',
        'om',
        ', a',
        's o',
        'ur',
        'li',
        'll',
        'ch',
        'had',
        'this',
        'e t',
        'g ',
        ' wh',
        'ere',
        ' co',
        'e o',
        'a ',
        'us',
        ' d',
        'ss',
        ' be',
        ' e',
        's a',
        'ma',
        'one',
        't t',
        'or ',
        'but',
        'el',
        'so',
        'l ',
        'e s',
        's,',
        'no',
        'ter',
        ' wa',
        'iv',
        'ho',
        'e a',
        ' r',
        'hat',
        's t',
        'ns',
        'ch ',
        'wh',
        'tr',
        'ut',
        '/',
        'have',
        'ly ',
        'ta',
        ' ha',
        ' on',
        'tha',
        '-',
        ' l',
        'ati',
        'en ',
        'pe',
        ' re',
        'there',
        'ass',
        'si',
        ' fo',
        'wa',
        'ec',
        'our',
        'who',
        'its',
        'z',
        'fo',
        'rs',
        'ot',
        'un',
        'im',
        'th ',
        'nc',
        'ate',
        'ver',
        'ad',
        ' we',
        'ly',
        'ee',
        ' n',
        'id',
        ' cl',
        'ac',
        'il',
        'rt',
        ' wi',
        'e, ',
        ' it',
        'whi',
        ' ma',
        'ge',
        'x',
        'e c',
        'men',
        '.com',
      ]);
    })
    .add('#compress', function() {
      compress('');
      compress('This is a small string');
      compress('foobar');
      compress('the end');
      compress('not-a-g00d-Exampl333');
      compress('Smaz is a simple compression library');
      compress('Nothing is more difficult, and therefore more precious, than to be able to decide');
      compress('this is an example of what works very well with smaz');
      compress('1000 numbers 2000 will 10 20 30 compress very little');
      compress('Nel mezzo del cammin di nostra vita, mi ritrovai in una selva oscura');
      compress('Mi illumino di immenso');
      compress("L'autore di questa libreria vive in Sicilia");
      compress('http://google.com');
      compress('http://programming.reddit.com');
      compress('http://github.com/antirez/smaz/tree/master');
    })
    .add('#decompress', function() {
      decompress(new Uint8Array([]));
      decompress(new Uint8Array([254, 84, 76, 56, 168, 61, 169, 152, 61, 191, 70]));
      decompress(new Uint8Array([216, 6, 90, 79]));
      decompress(new Uint8Array([1, 167, 60]));
      decompress(new Uint8Array([132, 200, 4, 200, 58, 255, 2, 48, 48, 24, 200, 254, 69, 241, 4, 45, 59, 22, 255, 3, 51, 51, 51]));
      decompress(new Uint8Array([254, 83, 169, 215, 56, 168, 61, 220, 59, 87, 160, 45, 59, 33, 165, 107, 201, 8, 90, 130, 12, 83]));
      decompress(new Uint8Array([254, 78, 218, 102, 99, 116, 45, 42, 11, 129, 44, 44, 131, 38, 22, 3, 148, 62, 206, 68, 11, 45, 42, 11, 59, 33, 28, 144, 163, 36, 199, 143, 96, 92, 25, 90, 87, 82, 164, 211, 230, 2]));
      decompress(new Uint8Array([155, 56, 168, 41, 2, 241, 4, 45, 59, 87, 32, 158, 135, 64, 42, 254, 107, 23, 224, 71, 145, 152, 235, 221, 10, 169, 215]));
      decompress(new Uint8Array([255, 4, 49, 48, 48, 48, 229, 38, 45, 92, 217, 0, 255, 4, 50, 48, 48, 48, 235, 152, 0, 255, 2, 49, 48, 0, 255, 2, 50, 48, 0, 255, 2, 51, 48, 160, 45, 59, 33, 165, 0, 224, 71, 151, 3, 3, 87]));
      decompress(new Uint8Array([254, 78, 174, 123, 2, 215, 215, 96, 106, 176, 28, 4, 45, 45, 105, 129, 229, 6, 77, 130, 0, 109, 47, 4, 36, 45, 8, 185, 47, 115, 109, 4, 8, 78, 0, 219, 162, 95, 22, 109, 162, 6, 10, 28, 150, 4]));
      decompress(new Uint8Array([254, 77, 8, 56, 152, 38, 45, 15, 96, 129, 56, 45, 243, 175]));
      decompress(new Uint8Array([255, 2, 76, 39, 4, 192, 42, 11, 129, 0, 254, 113, 38, 54, 196, 201, 8, 90, 33, 114, 162, 109, 182, 11, 105, 254, 83, 131, 233, 8, 4]));
      decompress(new Uint8Array([66, 58, 6, 6, 58, 87, 244]));
      decompress(new Uint8Array([66, 59, 115, 58, 130, 45, 45, 70, 110, 33, 24, 129, 3, 244]));
      decompress(new Uint8Array([66, 58, 47, 18, 38, 90, 244, 193, 26, 74, 33, 215, 193, 10, 169, 215, 193, 191, 228, 193, 169, 77, 27]));
    })
    .on('cycle', function(event) {
      console.log(String(event.target));
    })
    .on('complete', function() {
      console.log('Fastest is ' + this.filter('fastest').map('name'));
    })
    .run({ 'async': false });
})();
